opal/corelib/kernel.rb

```diff
@@ -108,16 +108,22 @@ def copy_singleton_methods(other)
       var i, name, names, length;
 
       if (other.hasOwnProperty('$$meta')) {
-        var other_singleton_class_proto = Opal.get_singleton_class(other).$$proto;
-        var self_singleton_class_proto = Opal.get_singleton_class(self).$$proto;
-        names = Object.getOwnPropertyNames(other_singleton_class_proto);
+        var other_singleton_class = Opal.get_singleton_class(other);
+        var self_singleton_class = Opal.get_singleton_class(self);
+        names = Object.getOwnPropertyNames(other_singleton_class.prototype);
 
         for (i = 0, length = names.length; i < length; i++) {
           name = names[i];
-          if (name.charAt(0) === '$') {
-            self_singleton_class_proto[name] = other_singleton_class_proto[name];
+          if (name.charAt(0) === '$' && name !== '$$id') {
+            self_singleton_class.prototype[name] = other_singleton_class.prototype[name];
           }
         }
+
+        self_singleton_class.$$const = Object.assign({}, other_singleton_class.$$const);
+        Object.setPrototypeOf(
+          self_singleton_class.prototype,
+          Object.getPrototypeOf(other_singleton_class.prototype)
```

Comment:
Yeah, that's because there's too much dynamic stuff in the RubySpec suite. It create anonymous classes, extends them in runtime, I guess that's the reason. For a static system of modules it seems to be faster.

Reply:
👍🏼 I need to test this on FF/mobile/… with more realistic code, just to be sure, anyway using prototypes makes everything 💯 times cleaner ✨